File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.java
class Solution {
    public int firstStableIndex(int[] nums, int k) {

        int n = nums.length;
        int[] suffix_min = new int[n];
        suffix_min[n - 1] = nums[n - 1];
        int maxx = Integer.MIN_VALUE;

        for (int i = n - 2; i >= 0; i--) {
            suffix_min[i] = Math.min(nums[i], suffix_min[i + 1]);
        }

        for (int i = 0; i < n; i++) {
            maxx = Math.max(maxx, nums[i]);
            int minn = suffix_min[i];

            if (maxx - minn <= k) {
                return i;
            }
        }

        return -1;
    }
}

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
# class Solution(object):
#     def firstStableIndex(self, nums, k):
#         for i in range (len(nums)):
#             maxx = max(nums[:i+1])
#             minn = min(nums[i:])
#             if abs(minn - maxx)<= k:
#                 return i
#         return -1

class Solution(object):
    def firstStableIndex(self, nums, k):

        n = len(nums)
        suffix_min = [0] * n
        suffix_min[-1] = nums[-1]
        maxx = float('-inf')

        for i in range(n - 2, -1, -1):
            suffix_min[i] = min(nums[i], suffix_min[i + 1])

        for i in range(n):
            maxx = max(maxx, nums[i])
            minn = suffix_min[i]
            if maxx - minn <= k:
                return i
                
        return -1